language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java

```diff
@@ -77,87 +87,47 @@ private CodeActionUtil() {
     /**
      * Get the top level node type at the cursor line.
      *
-     * @param context {@link LSContext}
+     * @param node node
      * @return {@link String}   Top level node type
      */
-    public static Optional<Pair<CodeActionNodeType, NonTerminalNode>> codeActionNodeType(CodeActionContext context) {
-        Position position = context.cursorPosition();
-        Optional<SyntaxTree> syntaxTree = context.workspace().syntaxTree(context.filePath());
-
-        if (syntaxTree.isEmpty()) {
-            return Optional.empty();
+    public static CodeActionNodeType codeActionNodeType(Node node) {
+        if (node == null) {
+            return CodeActionNodeType.NONE;
         }
-
-        ModulePartNode modulePartNode = syntaxTree.get().rootNode();
-        int cursorPosOffset = syntaxTree.get().textDocument().textPositionFrom(LinePosition.from(position.getLine(),
-                position.getCharacter()));
-
-        List<NonTerminalNode> members = modulePartNode.members().stream().collect(Collectors.toList());
-        modulePartNode.imports().forEach(members::add);
-        for (NonTerminalNode member : members) {
-            boolean isWithinStartSegment = isWithinStartCodeSegment(member, cursorPosOffset);
-            boolean isWithinBody = isWithinBody(member, cursorPosOffset);
-            if (!isWithinStartSegment && !isWithinBody) {
-                continue;
-            }
-
-            if (member.kind() == SyntaxKind.SERVICE_DECLARATION) {
-                if (isWithinStartSegment) {
-                    // Cursor on the service
-                    return Optional.of(new ImmutablePair<>(CodeActionNodeType.SERVICE, member));
+        switch (node.kind()) {
+            case SERVICE_DECLARATION:
```

Comment:
What would be the consequence if we try to depend on the node kind instance of code action node kind?

Reply:
I tried to implement it that way, issue is we cannot distinguish nodes in different contexts with node syntax-kind alone. For instance when you have a function definition syntax-kind, there's no clue whether it is a resource-function or a top-level function.